staking/README.md

```diff
@@ -0,0 +1,541 @@
+setUpdates := helper.GetUpdatedValidators(
+	&currentValidatorSet, // pointer to current validator set -- UpdateValidators will modify it
+	allValidators,        // All validators
+	ackCount,             // ack count
+)
+
+if len(setUpdates) > 0 {
+	// create new validator set
+	if err := currentValidatorSet.UpdateWithChangeSet(setUpdates); err != nil {
+		// return with nothing
+		logger.Error("Unable to update current validator set", "Error", err)
+		return abci.ResponseEndBlock{}
+	}
+
+	// validator set change
+	logger.Debug("[ENDBLOCK] Updated current validator set", "proposer", currentValidatorSet.GetProposer())
+
+	// save set in store
+	if err := app.StakingKeeper.UpdateValidatorSetInStore(ctx, currentValidatorSet); err != nil {
+		// return with nothing
+		logger.Error("Unable to update current validator set in state", "Error", err)
+		return abci.ResponseEndBlock{}
+	}
+
+// more code
+	}
+```
+
+## How to propose a MsgValidatorJoin transaction
+
+The `bridge` service in an existing validator's heimdall process polls for `Staked` event periodically and generates and broadcasts the transaction once it detects and parses the event. An existing validator on the network can also leverage the CLI to send the transaction:
+
+```
+heimdallcli tx staking validator-join --proposer <PROPOSER_ADDRESS> --tx-hash <ETH_TX_HASH> --signer-pubkey <PUB_KEY> --staked-amount <STAKED_AMOUNT> --activation-epoch <ACTIVATION_EPOCH>
+```
+
+Or the REST server :
+
+```
+curl -X POST "localhost:1317/staking/validator-join?from=<PROPOSER_ADDRESS>&tx-hash=<ETH_TX_HASH>&signer-pubkey=<PUB_KEY>&staked-amount=<STAKED_AMOUNT>&activation-epoch=<ACTIVATION_EPOCH>"
+```
+
+## How does an existing validator exit the network
+
+If a validator wishes to exit the network and unbond its stake, it invokes the `unstake` function on the `StakeManager` contract on L1, which emits an `UnstakeInit` event:
+
+```
+/// @param user address of the validator.
+/// @param validatorId unique integer to identify a validator.
+/// @param nonce to synchronize the events in heimdall.
+/// @param deactivationEpoch last epoch for validator.
+/// @param amount staking amount.
+event UnstakeInit(
+    address indexed user,
+    uint256 indexed validatorId,
+    uint256 nonce,
+    uint256 deactivationEpoch,
+    uint256 indexed amount
+);
+```
+
+An existing validator on Heimdall catches and parses this event and sends a `MsgValidatorExit` transaction, which is represented by the data structure:
+
+```
+type MsgValidatorExit struct {
+	From              hmTypes.HeimdallAddress `json:"from"`
+	ID                hmTypes.ValidatorID     `json:"id"`
+	DeactivationEpoch uint64                  `json:"deactivationEpoch"`
+	TxHash            hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex          uint64                  `json:"log_index"`
+	BlockNumber       uint64                  `json:"block_number"`
+	Nonce             uint64                  `json:"nonce"`
+}
+```
+where ,
```

Comment:
i'd remove the comma

Reply:
Done: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)